### PR TITLE
Fix issue with global variable as default value for class field and function parameter

### LINKED ISSUE #40769: [Bug]: Incorrect result when using global variable as a default value for class field and parameter
### Description

```ballerina
import ballerina/io;

final int recI = 111222;

class ClassWithDefaultsWithoutInitFunc {
    int i = recI;
}

function value(int k = recI) returns int {
    return k;
}

public function main() {
    ClassWithDefaultsWithoutInitFunc c1 = new;
    int k = c1.i;
    io:println(k); // prints 0 should be 111222
}
```

The provided code demonstrates an issue where a global variable, 'recI', is used as the default value for both a class field and a function parameter.  However, the expected behavior of initializing these values with the global variable is not met. Instead, the default value is incorrectly set to 0, leading to incorrect output when running the code. 

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #40769 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java [hunk 1]
```diff
@@ -64,6 +64,11 @@ public void testTypeRef() {
         BRunUtil.invoke(compileResult, "testTypeRefInClass");
     }
 
+    @Test
+    public void testClassWithDefaultsWithoutInitFunc() {
+        BRunUtil.invoke(compileResult, "testClassWithDefaultsWithoutInitFunc");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java:69]: Please add tests with module-level and local object constructors too.

And an actual closure scenario for a local object constructor.

Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java:69]: Is this name correct? Shouldn't it reflect the reference to the module-level variable?

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal [hunk 2]
```diff
@@ -70,6 +70,21 @@ class Per {
...
+class ClassWithDefaultsWithoutInitFunc {
+    int i = classI;
+}
+
+function value(int k = classI) returns int {
+    return k;
+}
+
+function testClassWithDefaultsWithoutInitFunc() {
+    ClassWithDefaultsWithoutInitFunc c = new();
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal:84]:
```suggestion
    ClassWithDefaultsWithoutInitFunc c = new;
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java (hunk 1 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java]: Changed to `testModuleVariableReferencingClass`

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal [hunk 3]
```diff
@@ -70,6 +70,33 @@ class Per {
...
+
+function value(int k = classI) returns int {
+    return k;
+}
+
+ModuleVariableReferencingClass c1 = new;
+
+function testModuleVariableReferencingClass() {
+    ModuleVariableReferencingClass c = new;
+    assertEquality(c.i, 111222);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal:87]: Need to swap values? First arg is meant to be expected and second actual?

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal [hunk 4]
```diff
@@ -70,6 +70,33 @@ class Per {
...
+    assertEquality(c1.i, 111222);
+}
+
+function testLocalObjectConstructorReferencingModuleVariable() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal:91]: Can we make the names uniform? They are both doing the same thing, just different constructs, right?

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal [hunk 5]
```diff
@@ -70,6 +70,33 @@ class Per {
...
+    var value = object {
+        int i = classI;
+        function init() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal:94]: Is the `init` method required for the test?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java [hunk 6]
```diff
@@ -64,6 +64,16 @@ public void testTypeRef() {
         BRunUtil.invoke(compileResult, "testTypeRefInClass");
     }
 
+    @Test
+    public void testModuleVariableReferencingClass() {
+        BRunUtil.invoke(compileResult, "testModuleVariableReferencingClass");
+    }
+
+    @Test
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java:72]: Is this the correct file for this test? Maybe both can go in `org.ballerinalang.test.object.ObjectTest`?

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal (hunk 3 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal]: Fixed.

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal (hunk 4 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal]: Updated as `testClassReferencingModuleVariable` and `testLocalObjectConstructorReferencingModuleVariable`

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal (hunk 5 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_class.bal]: Removed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java (hunk 6 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/klass/ClassTest.java]: Moved to `ObjectTest`

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal [hunk 7]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2023 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal:1]: Need to use the LLC header, without "All Rights Reserved".

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal:1]: Can't this go in an existing bal file? Can reduce a compilation.

File names use underscores if we are keeping this.

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+function testLocalObjectConstructorReferencingModuleVariable() {
+    var value = object {
+        int i = classI;
+    };
+    assertEquality(111222, value.i);
+}
+
+function assertEquality(any|error expected, any|error actual) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal:42]: Aren't you only passing `anydata` atm? Please update the function accordingly.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java [hunk 9]
```diff
@@ -47,11 +47,14 @@ public class ObjectTest {
 
     private CompileResult checkInInitializerResult;
     private CompileResult checkFunctionReferencesResult;
+    private CompileResult checkModuleVariableReferencesResult;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java:50]: Isn't this only when used as default values and therefore, closures? Please update the file, test, etc. names accordingly.

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal:1]: Moved to `object-with-defaultable-field.bal` file

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal (hunk 7 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal]: Removed this file

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal (hunk 8 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-referencing-module-variable.bal]: Yes. Updated.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java (hunk 9 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java]: Updated the names of test functions

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java [hunk 10]
```diff
@@ -47,11 +47,14 @@ public class ObjectTest {
 
     private CompileResult checkInInitializerResult;
     private CompileResult checkFunctionReferencesResult;
+    private CompileResult checkObjectWithDefaultableFieldsResult;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java:50]: Field with default value. Or just object with default values.

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal [hunk 11]
```diff
@@ -18,3 +18,43 @@ class Person {
...
+function testObjectWithModuleDefaultValue() {
+    var value = object {
+        int i = classI;
+    };
+    assertEquality(111222, value.i);
+}
+
+const ASSERTION_ERROR_REASON = "AssertionError";
+
+function assertEquality(any expected, any actual) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal:49]: Why any and not anydata?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java (hunk 10 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java]: change as `checkObjectWithDefaultValuesResult`

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java [hunk 12]
```diff
@@ -941,6 +942,16 @@ public void testNonPublicSymbolsWarningInServiceClass() {
         Assert.assertEquals(result.getDiagnostics().length, 0);
     }
 
+    @Test
+    public void testClassWithModuleDefaultValue() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java:946]: What does module default value mean? 

```suggestion
    public void testClassWithModuleLevelVarAsDefaultValue() {
```

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java:946]: Updated. I am trying to specify this test case using a module variable as a default value.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java [hunk 13]
```diff
@@ -941,6 +942,16 @@ public void testNonPublicSymbolsWarningInServiceClass() {
...
     }
 
+    @Test
+    public void testClassWithModuleLevelVarAsDefaultValue() {
+        BRunUtil.invoke(checkObjectWithDefaultValuesResult, "testClassWithModuleLevelVarAsDefaultValue");
+    }
+
+    @Test
+    public void testObjectWithModuleLevelVarAsDefaultValue() {
+        BRunUtil.invoke(checkObjectWithDefaultValuesResult, "testObjectWithModuleLevelVarAsDefaultValue");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/object/ObjectTest.java:952]:
```suggestion
    public void testObjectConstructorWithModuleLevelVarAsDefaultValue() {
        BRunUtil.invoke(checkObjectWithDefaultValuesResult, "testObjectConstructorWithModuleLevelVarAsDefaultValue");
```

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal [hunk 14]
```diff
@@ -18,3 +18,43 @@ class Person {
...
+
+ModuleVariableReferencingClass c1 = new;
+
+function testClassWithModuleLevelVarAsDefaultValue() {
+    ModuleVariableReferencingClass c = new;
+    assertEquality(111222, c.i);
+    assertEquality(111222, c1.i);
+}
+
+function testObjectWithModuleLevelVarAsDefaultValue() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal:40]:
```suggestion
function testObjectConstructorWithModuleLevelVarAsDefaultValue() {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal [hunk 15]
```diff
@@ -18,3 +18,43 @@ class Person {
...
+    var value = object {
+        int i = classI;
+    };
+    assertEquality(111222, value.i);
+}
+
+const ASSERTION_ERROR_REASON = "AssertionError";
+
+function assertEquality(anydata expected, anydata actual) {
+    if expected is anydata && actual is anydata && expected == actual {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal:50]: Don't you get warnings for the checks now? Always true?

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal:50]: Please update the logic, removing unnecessary checks.

Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal]: didn't get the warning.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/object/object-with-defaultable-field.bal]: Hint rather. Didn't you get `unnecessary condition: expression will always evaluate to 'true'`?

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**